R/import_config.R

```diff
@@ -0,0 +1,173 @@
+#' Get dettl config from path.
+#'
+#' Get the config containing information about location of functions containing
+#' import code.
+#'
+#' Sources any references to functions ready for use. Usea a default
+#' configuration in the result that any fields are missing. This will use the
+#' field name as the expected file name and function anme. Required fields are
+#' extract, transform, test_queries, test and load.
+#'
+#' @param path Path to directory containing config.
+#'
+#' @return The config as a dettl_config object.
+#'
+#' @keywords internal
+#'
+read_config <- function(path) {
+  filename <- file.path(path, "dettl.yml")
+  assert_file_exists(path, name = "Report working directory")
+  assert_file_exists(filename, name = "Dettl configuration")
+  info <- yaml_read(filename)
+
+  ## If certain fields don't exist in the config then add defaults
+  info <- validate_load(info)
+  function_fields <- c(
+    "extract",
+    "transform",
+    "load"
+  )
+  info <- add_missing_function_fields(info, function_fields)
+  required <- c(function_fields, "sources")
+  optional <- c("rewrite_keys")
+  check_fields(info, filename, required, optional)
+  if (info$load$default) {
+    info$rewrite_keys <- ForeignKeyConstraints$new(info$rewrite_keys)
+  }
+  env <- load_sources(info$sources, path)
+  info <- read_function_fields(function_fields, info, env)
+  info$name <- basename(normalizePath(path))
+  info$path <- path
+  class(info) <- "dettl_config"
+  info
+}
+
+validate_load <- function(info) {
+  default <- !is.null(info$load$default) && (info$load$default || tolower(info$load$default) == "true")
+  if (xor(default, !is.null(info$load$func))) {
+    info$load$default <- default
+  } else {
+    stop(sprintf("Load stage must specify a load function OR use the default load function. Got default %s and NULL func %s.",
+                 default, is.null(info$load$func)))
+  }
+  info
+}
+
+
+#' Add any missing required fields to the config using defaults where available.
+#'
+#' This defaults to using the field name for the function name.
+#' For the 'test' field also sets the verification queries function default and
+#' the test file default.
+#'
+#' @param info The config loaded from file.
+#' @param fields Collection of fields to set defaults for.
+#'
+#' @keywords internal
+#'
+add_missing_function_fields <- function(info, fields) {
+  for (field_name in fields) {
+    info <- set_missing_values(info, field_name)
+  }
+  info
+}
+
+set_missing_values <- function(info, field_name) {
+  ## Func can be empty for load field when running default load
+  if (field_name != "load") {
+    if (is.null(info[[field_name]]$func) || is.na(info[[field_name]]$func)) {
+      info[[field_name]]$func <- field_name
+    }
+  }
+  missing_verification_queries <- field_name == "load" &&
+    (is.null(info[[field_name]]$verification_queries) ||
```

Comment:
If this is coming from yaml is the `is.na` part needed?  If so, then pull it into a utility function (`is_missing <- function(x) is.null(x) || is.na(x)`) and use that to make this easier to read - but be aware it will only work sensibly if the value is a scalar

Reply:
You're right not needed. I've removed that.